tempesta_fw/ss_skb.c

```diff
@@ -313,8 +368,17 @@ __split_linear_data(struct sk_buff *skb, char *pspt, int len, TfwStr *it)
 	 */
 	if (unlikely((len < 0) && (tail_len == -len))) {
 		ss_skb_put(skb, len);
-		if (skb_is_nonlinear(skb))
+		if (skb_is_nonlinear(skb)) {
 			it->ptr = skb_frag_address(&skb_shinfo(skb)->frags[0]);
+			it->skb = pskb ? pskb : skb;
+		} else {
+			struct sk_buff *dskb = __next_skb_frag(skb, pskb);
+			if (dskb)
+				it->skb = pskb ? pskb : skb;
```

Comment:
I believe that is the correct behaviour. Suppose we store an skb fragment in `it->ptr`, and then do another modification that involves that skb. The modification may require that a new skb is created, as in `__extend_pgfrags()`. Then the new skb would be attached to the current skb as it's skb fragment which is wrong because the current skb in this case is an skb fragment itself. Skb fragments don't have fragments in the Linux kernel. Parent skb is the real skb, the others are skb fragments.